### DISCUSSION TITLE
Adapter Maintainers: Upgrading to dbt-core `v1.8.0`

### DISCUSSION THREAD

**Discussion by dataders** [dbt-labs/dbt-core#9798] [Community]
## Overview 

This discussion is for communicating to adapter maintainers the scope of work needed to make use of the changes coming via dbt-core 1.8.0. If you have questions and concerns, please ask them here for posterity.

Please consider this a living document between now and the date of final release. If there's something missing, please comment below!

<details><summary>

### Loom video overview (12 min)

</summary>

TBD

</details>

### release timeline

The below table gives the milestones between up to and including the final release. It will be updated with each subsequent release.

- [x] #9780
- [x] #9781
- [x] #9782
- [x] #9783

<details><summary>

### prior maintainer upgrade guides

</summary>

- #8307
- #7958
- #7213
- #6624
- #6011
- #5468

</details>

### Example Diffs

- <https://github.com/dbt-labs/dbt-redshift/compare/v1.7.0...v1.8.0b1>
- <https://github.com/dbt-labs/dbt-bigquery/compare/v1.7.0...v1.8.0b1>
- <https://github.com/dbt-labs/dbt-snowflake/compare/v1.7.0...v1.8.0b1>
- <https://github.com/dbt-labs/dbt-spark/compare/v1.7.0...v1.8.0b1>

## TL;DR

This upgrade is a big deal, and should be the last time in a while that dbt Labs asks maintainers for this amount of work. There's two logical pieces of work here, that, in theory, aren't especially complex:

- decoupling dependency on dbt-core
- supporting the new unit testing feature for end users

### Context

- [dbt Core Roadmap Update 11/23: Adapters & Artifacts](https://github.com/dbt-labs/dbt-core/blob/main/docs/roadmap/2023-11-dbt-tng.md#adapters--artifacts)
- [dbt-core#9171](https://github.com/dbt-labs/dbt-core/discussions/9171)
- [dbt-adapters#87](https://github.com/dbt-labs/dbt-adapters/discussions/87)

#### Excerpt from #9171

> The goals of the work are to do the following:
>
> - Dramatically lower the effort required to maintain an adapter plugin for dbt-core
> - Enable users to install the latest dbt-core version, even if their dbt-* adapter has not had a corresponding release

### How to implement

Everything you need to know is in [dbt-adapters#87](https://github.com/dbt-labs/dbt-adapters/discussions/87). If you have a question or concern, please ask it there.

## Unit Testing

Unit testing is a long-awaited feature that is finally coming to dbt. It is a powerful feature that will allow users to test their models in isolation from the rest of their project. This should not require a great deal of work on the part of the adapter maintainer, but, more importantly, how dbt is used changes, so it's important that we have test coverage for these new scenarios.s

### support for `--empty` flag to enable "dry run" mode

- [dbt-core#8980](https://github.com/dbt-labs/dbt-core/issues/8980): original issue
- [dbt-core#8971](https://github.com/dbt-labs/dbt-core/pull/8971): Core PR
- [dbt-redshift#666](https://github.com/dbt-labs/dbt-redshift/issues/666): Example implementation

#### How to implement `--empty` support

[dbt-core#8971](https://github.com/dbt-labs/dbt-core/pull/8971) added a new BaseRelation method: [`render_limited()`](https://github.com/dbt-labs/dbt-core/blob/7967be7bb373a3c737196bc0ebbe31ef6f4ed354/core/dbt/adapters/base/relation.py#L198-L205). Effectively, this method will wrap a model's `SELECT` statement into a subquery

```sql
(select * from {rendered} limit {self.limit}) _dbt_limit_subq

-- if limit is 0 (`where false` for cost saving)
(select * from {rendered} where false limit 0) _dbt_limit_subq
```

If your data platform supports `LIMIT` clauses, you have no work to do. However, some SQL dialects (e.g. T-SQL) do not support `LIMIT` clauses. In this case, you will need to implement `render_limited()` for your adapter.

#### `--empty` Tests

`dbt.tests.adapter.empty.test_empty.BaseTestEmpty`

### macros `CAST` and `SAFE_CAST` support

dbt does a vast amount of type casting behind the scenes such as:

- data warehouse data types <> Python
- `agate` (our current csv reader) <> Python
- `YAML` <> Python

In the case of unit testing, we needed to extend further the `yaml`<>`python` casting to allow for more specific definition of mock data.

theoretically this should be a no-op for most adapters, but it's worth checking to make sure that the `CAST` and `SAFE_CAST` macros are supported in your adapter. For example, dbt-spark now has a `safe_cast` that it did not before ([dbt-spark#files](https://github.com/dbt-labs/dbt-spark/pull/976/files)).

#### Tests within `dbt.tests.adapter.unit_testing`

For unit testing, there are a handful (3) of functional tests worth implementing as an adapter to ensure both baseline functionality and expected behavior when mocking inputs with various types:

##### `test_types.BaseUnitTestingTypes`

- this test sets up a unit test on a test adapter that mocks out its input with varying data types ([base implementation](https://github.com/dbt-labs/dbt-adapters/blob/main/dbt-tests-adapter/dbt/tests/adapter/unit_testing/test_types.py))
- implementing this should be a matter of overriding the `data_types` fixture ([example](https://github.com/dbt-labs/dbt-bigquery/blob/main/tests/functional/adapter/unit_testing/test_unit_testing.py#L9))
- `data_types` is a list of (`sql_value`, `yaml_value`) where `sql_value` should be a literal in the upstream 'real' input, and `yaml_value` is what the value looks like in yaml when it is being mocked out by the user
- under the hood, the `unit` materialization makes use of `safe_cast` to cast the user-provided yaml value to the expected input type while building fixture CTEs
    - `safe_cast` defers to `cast` if the adapter does not support safe casting to a particular type (e.g. [snowflake's `safe_cast` does not support variant](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/include/snowflake/macros/utils/safe_cast.sql#L7))
    - so: the adapter's implementation of `safe_cast` and `cast` may need to be extended as appropriate to support a fuller range of inputs that can be expected from the user in the context of specifying unit test input fixtures.

##### `test_case_insensitivity.BaseUnitTestCaseInsensivity`

> [!NOTE]
there is a misspelling, `Insensivity` not `Insensitivity`

- this is more of a baseline behavior test, that ensures input fixtures can specify column names in a case-insensitive manner ([base implementation](https://github.com/dbt-labs/dbt-adapters/blob/main/dbt-tests-adapter/dbt/tests/adapter/unit_testing/test_case_insensitivity.py))
- should be handled by the default implementation and not require anything beyond a passthrough implementation ([example](https://github.com/dbt-labs/dbt-bigquery/pull/1031/files#diff-fa16d6a4b96751c43394815126f09d409c56cc89baff1a089af16c15e55118baR59))
    - unless they are overwriting the unit materialization or [`get_fixture_sql`](https://github.com/dbt-labs/dbt-adapters/blob/35bd3629c390cf87a0e52d999679cc5e33f36c8f/dbt/include/global_project/macros/unit_test_sql/get_fixture_sql.sql#L1), which I'd advise against as those provide the main adapter framework/entrypoints of the unit testing functionality and our 1p implementations have not had to.

##### `test_invalid_input.BaseUnitTestInvalidInput`

- another baseline test that ensures the appropriate error & messaging is raised by the adapter if a user specifies an invalid column name in an input fixture ([base implementation](https://github.com/dbt-labs/dbt-adapters/blob/main/dbt-tests-adapter/dbt/tests/adapter/unit_testing/test_invalid_input.py))
- should be handled by the default implementation and not require anything beyond a passthrough implementation ([example](https://github.com/dbt-labs/dbt-bigquery/pull/1031/files#diff-fa16d6a4b96751c43394815126f09d409c56cc89baff1a089af16c15e55118baR63))
    - unless they are overwriting the `format_row` macro which is what [provides this functionality in the default implementation](https://github.com/dbt-labs/dbt-adapters/blob/35bd3629c390cf87a0e52d999679cc5e33f36c8f/dbt/include/global_project/macros/unit_test_sql/get_fixture_sql.sql#L75-L78).

### Additional Tests

> [!IMPORTANT]
These are new tests introduced into the adapter zone that you should have in your adapter.

TBD

### Materialized Views Refactor

When the `1.7` upgrade guide was originally published (October 2023), the "Materialized Views Refactor" was stubbed out. In December it was more fully fleshed out.

If your warehouse supports materialized views, you should check it out. Even if you don't the changes implemented represent a vision of the future for how materializations are handled in dbt.

**Comment by benc-db** [comment]
Asking here per request:

As this uncouples core and adapters, that seems like a pretty big shift.  Why is this not version 2.0.0?  Or, if for internal reasons it must be 1.8.0, are adapters free to make their release that is compatible with this into version 2.0.0?  I want to signal to customers that they can't just take latest without making the change, particularly in environments where they get latest by default, like in Databricks Workflows.

**Reply by dataders** [reply]
>  Why is this not version `2.0.0`?
The SemVer guarantee is to dbt end users. In a perfect world, we would have also made the same guarantee for adapter maintainers, but didn't. So `2.0.0` would mean backward-incompatible changes for end users, which, provided that adapter maintainers respond appropriately, has yet to (and will not) happen before `v2.0`.

Another way to think about this is that we're doing what ideally would have been done with `v1.0` if we wanted to ensure forwards compatibility for adapters

> are adapters free to make their release that is compatible with this into version 2.0.0?

Yes! This is the lowest-hanging, biggest-impact change IMHO. dbt-databricks can now follow SemVer itself. Make whatever version you like! I can see your argument for `2.0.0`.

Ultimately I'd consider end users. If they have their dependency defined as `dbt-databricks~=1.7.0`, would we want them to catch this update? My feeling is yes from an end-user standpoint, given that there are no backwards-incompatible changes.

**Reply by benc-db** [reply]
No, we wouldn't, because they would also need to install dbt-core, which they are most likely not doing today; they would get a broken experience with no indicator as to why it's broken.

**Comment by dataders** [comment]
using a comment to track progress across all (known) adapters

<details><summary>

### `1.8` upgrade issues

</summary>

- [ ] https://github.com/databricks/dbt-databricks/issues/619
- [ ] https://github.com/dbt-athena/dbt-athena/issues/610
- [ ] https://github.com/starburstdata/dbt-trino/issues/391
- [ ] https://github.com/ClickHouse/dbt-clickhouse/issues/260
- [ ] https://github.com/duckdb/dbt-duckdb/issues/368
- [ ] https://github.com/MaterializeInc/materialize/issues/26226
- [ ] https://github.com/oracle/dbt-oracle/issues/138
- [ ] https://github.com/codeforkjeff/dbt-sqlite/issues/52
- [ ] https://github.com/microsoft/dbt-synapse/issues/227
- [ ] https://github.com/microsoft/dbt-fabric/issues/146
- [ ] https://github.com/Teradata/dbt-teradata/issues/151
- [ ] https://github.com/dremio/dbt-dremio/issues/222
- [ ] https://github.com/dbt-msft/dbt-sqlserver/issues/492
- [ ] https://github.com/aws-samples/dbt-glue/issues/364
- [ ] https://github.com/tglunde/dbt-exasol/issues/120
- [ ] https://github.com/firebolt-db/dbt-firebolt/issues/124
- [ ] https://github.com/memsql/dbt-singlestore/issues/23
- [ ] https://github.com/chop-analytics/dbt-netezza/issues/25
- [ ] https://github.com/databendcloud/dbt-databend/issues/6
- [ ] https://github.com/vertica/dbt-vertica/issues/131
- [ ] https://github.com/markporoshin/dbt-greenplum/issues/34
- [ ] https://github.com/cloudera/dbt-hive/issues/134
- [ ] https://github.com/aurany/dbt-ibmdb2/issues/56
- [ ] https://github.com/cloudera/dbt-impala/issues/194
- [ ] https://github.com/iomete/dbt-iomete/issues/11
- [ ] https://github.com/mindsdb/dbt-mindsdb/issues/32
- [ ] https://github.com/dbeatty10/dbt-mysql/issues/170
- [ ] https://github.com/rockset/dbt-rockset/issues/40
- [ ] https://github.com/pingcap/dbt-tidb/issues/55
- [ ] https://github.com/fal-ai/fal/issues/139
- [ ] https://github.com/inferlabs/dbt-infer/issues/10
- [ ] https://github.com/getindata/dbt-flink-adapter/issues/79
- [ ] https://github.com/tanyshak/dbt-upsolver/issues/6

</details>

**Comment by benc-db** [comment]
How do we decide the full spectrum of SQL types and yaml types to put in the type tests?  Is it basically, here are all the SQL types that my database supports, and here are all the ways I expect users might want to specify it?  Is there any standardized way to communicate this mapping to users?

**Reply by dataders** [reply]
great question! I was just told that this doc exists that shows some current examples:
[dbt Docs: Platform-specific Data types](https://docs.getdbt.com/reference/resource-properties/data-types)

sidebar: we should also document the common Databricks types

**Comment by prdpsvs** [comment]
@dataders , test_unit_testing test is generating a nested CTE, which is not supported in Fabric yet.
![image](https://github.com/dbt-labs/dbt-core/assets/19339281/a971fae4-4161-4b48-bf0f-9a076808cbd9)
any thoughts?

**Comment by prdpsvs** [comment]
Also, I added render_limited() as limit key word is not supported by t-sql. It looks like the default implementation from dbt-core is picked up instead of using the adapter implementation.

![image](https://github.com/dbt-labs/dbt-core/assets/19339281/f6bb3448-17ff-4485-888c-2c72303bcb7e)

**Reply by dataders** [reply]
So you added a [`.render_limited()`](https://github.com/dbt-labs/dbt-core/blob/7967be7bb373a3c737196bc0ebbe31ef6f4ed354/core/dbt/adapters/base/relation.py#L198-L205) that works with TSQL. but your problem is that your override is not being used when the tests run? Or are you saying that the test is defined to use a nested CTE?

**Reply by prdpsvs** [reply]
I added render_limited in the adapter class. When I run BaseTestEmpty test, I see that dbt-core implementation is being used but not the adapter implementation (below method)

```
def render_limited(self) -> str:
        rendered = self.render()
        if self.limit is None:
            return rendered
        elif self.limit == 0:
            return f"(select * from {rendered} where 1=0) _dbt_top_subq"
        else:
            return f"(select TOP {self.limit} * from {rendered}) _dbt_top_subq"
```

**Reply by dataders** [reply]
eek. I'll flag this to the adapters team. If you could link to a failing test in a draft PR, that'd be ideal, but I'll let them know regardless

**Reply by prdpsvs** [reply]
The CTE problem is with test_unit_testing.py test. The below schema.yml generates nested cte to compare the data types

```
test_my_model_yml = """
unit_tests:
  - name: test_my_model
    model: my_model
    given:
      - input: ref('my_upstream_model')
        rows:
          - {{ tested_column: {yaml_value} }}
    expect:
      rows:
        - {{ tested_column: {yaml_value} }}
"""
```

**Reply by dataders** [reply]
hey @prdpsvs -- you were right to flag this. the test case would never pass because it won't use your `.render_limited()`

From an engineer
> The test `test_render_limited()` in `tests/unit/test_relation.py` ([source](https://github.com/dbt-labs/dbt-adapters/blob/main/tests/unit/test_relation.py#L59-L68)) uses a `BaseRelation` not `FabricRelation`. Therefore the adapter's override of `.render_limited()` will never be used

```py
def test_render_limited(limit, expected_result):
    my_relation = BaseRelation.create(
        database="test_database",
        schema="test_schema",
        identifier="test_identifier",
        limit=limit,
    )
    actual_result = my_relation.render_limited()
    assert actual_result == expected_result
    assert str(my_relation) == expected_result
```

**Reply by prdpsvs** [reply]
@dataders , I added fabric relation https://github.com/microsoft/dbt-fabric/actions/runs/8700522952/job/23860851994#step:6:282
However, the render_limited from adapter implementation is not used. Please see the integration test failure.

**Reply by dataders** [reply]
you're almost there! I think you have to add a line to the [FabricAdapter dataclass](https://github.com/microsoft/dbt-fabric/blob/0b3a9065dbb8dabe4aa37fb8e48cdfe622682af1/dbt/adapters/fabric/fabric_adapter.py#L26-L29). See [SnowflakeAdapter for an example](https://github.com/dbt-labs/dbt-snowflake/blob/8b75a53864e1fd94e902657963ec8486b3e5dc53/dbt/adapters/snowflake/impl.py#L37-L40)

```py
Relation = FabricRelation
```

**Comment by s-kethiri** [comment]
Hi. Curious to know if anyone successfully implemented dbt-core (v1.8.6) unit testing with dbt-athena (v1.8.4) for iceberg tables. I tried running the unit test, but the test is attempting to create a hive table instead.
